k8sutils/pkg/source/source.go

```diff
@@ -123,3 +117,99 @@ func GetClientObjectFromSource(ctx context.Context, kubeClient client.Client, so
 
 	return obj, nil
 }
+
+func HandleInstrumentationConfigDataStreamsLabels(ctx context.Context,
+	workloadSources *odigosv1.WorkloadSources, ic *odigosv1.InstrumentationConfig) bool {
+	logger := log.FromContext(ctx)
+
+	workloadSource := workloadSources.Workload
+	namespaceSource := workloadSources.Namespace
+
+	// Case 1: no workload source and no namespace source — nothing to do
+	if workloadSource == nil && namespaceSource == nil {
+		logger.Info("No workload source and no namespace source found, skipping")
+		return false
+	}
+
+	// Extract labels from both sources (may be nil)
+	var workloadLabels map[string]string
+	if workloadSource != nil {
+		workloadLabels = getSourceDataStreamsLabels(workloadSource)
+	} else {
+		workloadLabels = map[string]string{}
+	}
+
+	var namespaceLabels map[string]string
+	if namespaceSource != nil {
+		namespaceLabels = getSourceDataStreamsLabels(namespaceSource)
+	} else {
+		namespaceLabels = map[string]string{}
+	}
+
+	// Start merging logic:
+	mergedLabels := make(map[string]string)
+
+	// Start from namespace labels (if any)
+	for k, v := range namespaceLabels {
+		mergedLabels[k] = v
+	}
+
+	// Add any workload keys not present in namespace
+	for k, v := range workloadLabels {
+		if _, exists := mergedLabels[k]; !exists {
+			mergedLabels[k] = v
+		}
+	}
+
+	// Apply overrides from workload source (force false where set)
+	for k, v := range workloadLabels {
+		if v == "false" {
+			mergedLabels[k] = "false"
+		}
+	}
+
+	// Apply merged labels into InstrumentationConfig, and return true if changed
+	return mergeInstrumentationConfigLabels(ic, mergedLabels)
+}
+
+func mergeInstrumentationConfigLabels(instConfig *odigosv1.InstrumentationConfig, desiredLabels map[string]string) (updated bool) {
+	if instConfig.Labels == nil {
+		instConfig.Labels = make(map[string]string)
+	}
+
+	// Add / update labels
+	for key, value := range desiredLabels {
+		if instConfig.Labels[key] != value {
+			instConfig.Labels[key] = value
+			updated = true
+		}
+	}
+
+	// Remove datastream labels not present in desiredLabels
+	for key := range instConfig.Labels {
+		if _, exists := desiredLabels[key]; !exists && isDataStreamLabel(key) {
+			delete(instConfig.Labels, key)
+			updated = true
+		}
+	}
```

Comment:
Right now IC labels are only for DS, so it makes sense for now.
But if we add non-DS labels in the future, this will remove all of them.

I think it's best not to remove any labels.
The `desiredLabels` contains DS labels with `false` values, I think that's the only kind of "removal" we would need for now.

```suggestion
        // do not remove
```